2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,394 @@
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will depend on https://www.nuget.org/packages/NuGet.Client.
+This module would be shipped in PowerShell 7 and on PSGallery supporting older
+versions of PowerShell.
+
+### Side-by-side with PowerShellGet
+
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+Script and module metadata will retain the same format as it exists with v2.
+
+### Local cache
+
+Instead of always connecting to PSGallery to perform an online search,
+`Find-PSResource` (see below) works against a local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+A system wide cache would require elevation or sudo to create and update preventing
+it from being useful.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "exportedFunctions": [
+    "Get-One",
+    "Set-Two"
+  ],
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources (approximately the number of unique packages
+currently published to PowerShellGallery.com) is ~700KB in compressed json form.
+
+The cache would have both latest stable and latest prerelease versions of resources.
+
+>[!NOTE]
+>Need to experiment if it makes sense to have a single cache file for all repositories
+>or a different cache per repository for size and perf reasons.
+>Perf tests will determine if the cache needs to be in binary form.
```

Comment:
If a cache is implemented then it needs to be able to be configured in such a way that allows SysAdmins flexibility in managing the cache across a large number of systems

Therefore a Local cache system should

- [ ] A Cache file should be able to be created even for repositories that may not support it via deeper package inspection for example allowing Fileshare bases repositories to be used with no API endpoint required. This would take time but would be beneficial in secured environments (High priority)

- [ ] Download once from the feed and push a single file to all clients (High priority)

- [ ] Allow admin editing of this cache to allow the admin to essentially have an allowable/disallowed packages list within the cache, enforcing a trusted packages model. This should be an optional feature with the default being that it contains all packages on a feed are allowed., Ideally this should be via some form of PowerShell command that can be run that would spit out a valid file that either only includes the allowed packages or marks a package as disallowed in the cache. (Medium priority)

- [ ] If a package which is marked as disallowed this should be installable but require an additional -Bypass type switch that emits an entry to Event Log as well as forwards an event to notify a SysAdmin (low priority)

- [ ] Implement a separate cache for each registered repository, the benefit being that a SysAdmin can push out updated cache file for each cache very easily to all clients at once 


- [ ] Allow the Admin to enforce that only the cache file is used and therefore that user can only install approved packages from public repositories (potentially mitigating the need for an in house package feed) and also stopping users installing modules that have not been internally approved. (High priority)

- [ ] Be reachable from a shared Network Drive for each different repository (mitigates the need to push out on slow networks) (Medium priority)

- [ ] Cache file should be mapped with a matching repository settings file encompassing these configurable settings (High priority)

- [ ] Cache Updating should be configurable in repository settings to allow it to only run on a specific schedule that invokes some form of scheduled task/job under the hood and also should be configurable as to disable the running of an update to the cache file whilst any operation reaches out to a repository to download from the repository, It should also be a configurable repository setting that then would allows all calls to that repository to update the cache file (High priority)

- [ ] Cache Updating should be able to be run on demand via a cmdlet allowing more flexibility around the updating of a cache file and should be able to be added as a switch to any Find/Install commands as to allow the user to decide to update on an individual run and not as part of every single run. (High priority)

Reply:
In the case that the repository doesn't support the request cache api, it'll fallback to the current behavior which is to make network requests to query for resources.  If the cache file format is in a well defined form, then presumably the admin could modify this to suit their needs.  This means a binary format would be out of the question (for now).  For your other items, I'm calling out enterprise management of local cache as out of scope for 3.0.

Reply:
@SteveL-MSFT I wrote up the following recommendations in the OneGet repository, which I will copy-paste here based on @SydneyhSmith suggestion (I did not realize that repository was did, so apologies for posting it there).

1. Align package installation to [ExecutionPolicyScope enum](https://github.com/PowerShell/PowerShell/blob/ceed73d7375cb98a199ceb19796bb2311b0b0002/src/System.Management.Automation/security/SecuritySupport.cs#L73-95)
2. Align package trust to [ExecutionPolicy enum](https://github.com/PowerShell/PowerShell/blob/ceed73d7375cb98a199ceb19796bb2311b0b0002/src/System.Management.Automation/security/SecuritySupport.cs#L31-L65)
3. Replace concept of "trust" and "cache" with unifiying concept of [merkle tree](https://hackernoon.com/merkle-trees-181cb4bc30b4) to separate data from the verification of the data
4. Solve [Feature Request: Ability to configure PSRepository on a system-wide level](https://github.com/PowerShell/PowerShellGet/issues/214)

Reply:
In the extreme, trust can be modeled using Microsoft Authenticode for code signing.  There can be multiple merkle trees, one of which is computed using a "twisted braid" of normal blockchain-style hashing algorithm and a Microsoft Authenticode value.

Not sure if Authenticode is portable to Linux?

Reply:
Authenticode isn't portable as it currently exists to Linux without taking some Windows OS code with it.  I like the concept of a merkle tree, but will defer it to post 3.0.  Policy for PSRepository registration should follow the [policy RFC](https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0041-Policy.md)